src/Workspaces/Core/Portable/SourceGeneratorTelemetry/SourceGeneratorTelemetryCollectorWorkspaceService.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Microsoft.CodeAnalysis.Internal.Log;
+
+namespace Microsoft.CodeAnalysis.SourceGeneratorTelemetry
+{
+    internal class SourceGeneratorTelemetryCollectorWorkspaceService : ISourceGeneratorTelemetryCollectorWorkspaceService
+    {
+        private record GeneratorTelemetryKey
+        {
+            public GeneratorTelemetryKey(ISourceGenerator generator)
+            {
+                Identity = new SourceGeneratorIdentity(generator);
+                FileVersion = FileVersionInfo.GetVersionInfo(generator.GetGeneratorType().Assembly.Location).FileVersion ?? "(null)";
+            }
+
+            // TODO: mark these 'required' when we have the attributes in place
```

Comment:
@333fred Do we already have the attributes to build against so I can start to use required in the Roslyn codebase?

Reply:
We don't have them in the codebase yet, and I don't think we can use the feature until we update to P6.